File: frontend/src/types/research.ts
// research domain types
import type { ManualLlmReviewView } from "./stock";

export interface ManualResearchRequestCreateRequest {
  symbol: string;
  question: string;
  trigger_source?: string;
  executor_kind: string;
  model_api_key_id?: number | null;
}

export interface ManualResearchRequestExecuteRequest {
  failover_enabled: boolean;
}

export interface ManualResearchRequestCompleteRequest {
  summary: string;
  review_verdict: string;
  risks: string[];
  disagreements: string[];
  decision_note?: string | null;
  citations: string[];
  answer?: string | null;
}

export interface ManualResearchRequestFailRequest {
  failure_reason: string;
}

export interface ManualResearchRequestRetryRequest {
  requested_by?: string | null;
}

export interface ManualResearchRequestView {
  id: number;
  request_key: string;
  recommendation_key: string;
  symbol: string;
  question: string;
  trigger_source: string;
  executor_kind: string;
  model_api_key_id?: number | null;
  status: string;
  status_note?: string | null;
  requested_at: string;
  started_at?: string | null;
  completed_at?: string | null;
  failed_at?: string | null;
  artifact_id?: string | null;
  failure_reason?: string | null;
  requested_by?: string | null;
  superseded_by_request_id?: number | null;
  stale_reason?: string | null;
  source_packet_hash: string;
  validation_artifact_id?: string | null;
  validation_manifest_id?: string | null;
  source_packet: string[];
  selected_key?: AnalysisKeySelectionView | null;
  attempted_keys: AnalysisAttemptView[];
  failover_used: boolean;
  manual_llm_review: ManualLlmReviewView;
}

export interface ManualResearchRequestListResponse {
  generated_at: string;
  counts: Record<string, number>;
  items: ManualResearchRequestView[];
}

export interface FollowUpAnalysisRequest {
  symbol: string;
  question: string;
  model_api_key_id?: number | null;
  failover_enabled: boolean;
}

export interface AnalysisAttemptView {
  key_id?: number | null;
  name: string;
  provider_name: string;
  model_name: string;
  status: string;
  error?: string | null;
}

export interface AnalysisKeySelectionView {
  id?: number | null;
  name: string;
  provider_name: string;
  model_name: string;
  base_url: string;
}

export interface FollowUpAnalysisResponse {
  symbol: string;
  question: string;
  request_id: number;
  request_key: string;
  status: string;
  executor_kind: string;
  status_note?: string | null;
  answer?: string | null;
  selected_key?: AnalysisKeySelectionView | null;
  failover_used: boolean;
  attempted_keys: AnalysisAttemptView[];
  manual_review_artifact_id?: string | null;
}

File: src/ashare_evidence/schemas/research.py
"""R e s e a r c h domain schemas."""
from __future__ import annotations

from datetime import date, datetime
from typing import Any

from pydantic import BaseModel, Field

from typing import TYPE_CHECKING
if TYPE_CHECKING:
    from .stock import ManualLlmReviewView


class AnalysisAttemptView(BaseModel):
    key_id: int | None = None
    name: str
    provider_name: str
    model_name: str
    status: str
    error: str | None = None


class AnalysisKeySelectionView(BaseModel):
    id: int | None = None
    name: str
    provider_name: str
    model_name: str
    base_url: str


class FollowUpAnalysisRequest(BaseModel):
    symbol: str
    question: str
    model_api_key_id: int | None = None
    failover_enabled: bool = True


class FollowUpAnalysisResponse(BaseModel):
    symbol: str
    question: str
    request_id: int
    request_key: str
    status: str
    executor_kind: str
    status_note: str | None = None
    answer: str | None = None
    selected_key: AnalysisKeySelectionView | None = None
    failover_used: bool = False
    attempted_keys: list[AnalysisAttemptView] = Field(default_factory=list)
    manual_review_artifact_id: str | None = None


class ManualResearchRequestCreateRequest(BaseModel):
    symbol: str
    question: str
    trigger_source: str = "manual_research_ui"
    executor_kind: str = "builtin_gpt"
    model_api_key_id: int | None = None


class ManualResearchRequestExecuteRequest(BaseModel):
    failover_enabled: bool = True


class ManualResearchRequestCompleteRequest(BaseModel):
    summary: str
    review_verdict: str
    risks: list[str] = Field(default_factory=list)
    disagreements: list[str] = Field(default_factory=list)
    decision_note: str | None = None
    citations: list[str] = Field(default_factory=list)
    answer: str | None = None


class ManualResearchRequestFailRequest(BaseModel):
    failure_reason: str


class ManualResearchRequestRetryRequest(BaseModel):
    requested_by: str | None = None


class ManualResearchRequestView(BaseModel):
    id: int
    request_key: str
    recommendation_key: str
    symbol: str
    question: str
    trigger_source: str
    executor_kind: str
    model_api_key_id: int | None = None
    status: str
    status_note: str | None = None
    requested_at: datetime
    started_at: datetime | None = None
    completed_at: datetime | None = None
    failed_at: datetime | None = None
    artifact_id: str | None = None
    failure_reason: str | None = None
    requested_by: str | None = None
    superseded_by_request_id: int | None = None
    stale_reason: str | None = None
    source_packet_hash: str
    validation_artifact_id: str | None = None
    validation_manifest_id: str | None = None
    source_packet: list[str] = Field(default_factory=list)
    selected_key: AnalysisKeySelectionView | None = None
    attempted_keys: list[AnalysisAttemptView] = Field(default_factory=list)
    failover_used: bool = False
    manual_llm_review: ManualLlmReviewView


class ManualResearchRequestListResponse(BaseModel):
    generated_at: datetime
    counts: dict[str, int] = Field(default_factory=dict)
    items: list[ManualResearchRequestView] = Field(default_factory=list)

